// portkey-ai/dist/tests/images/openai.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
const dotenv_1 = require("dotenv");
const portkey_ai_1 = require("portkey-ai");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
(0, dotenv_1.config)({ override: true });
const client = new portkey_ai_1.Portkey({
    apiKey: (_a = process.env["PORTKEY_API_KEY"]) !== null && _a !== void 0 ? _a : "",
    virtualKey: (_b = process.env["OPENAI_VIRTUAL_KEY"]) !== null && _b !== void 0 ? _b : ""
});
describe("Openai Images APIs", () => {
    test("generate: only required params", () => __awaiter(void 0, void 0, void 0, function* () {
        const response = yield client.images.generate({
            prompt: "A cute baby sea otter",
        });
        expect(response).toBeDefined();
        expect(response.data).toBeDefined();
        expect(response.data.length).toBeGreaterThan(0);
    }), 120000);
    test("generate: only required params with model", () => __awaiter(void 0, void 0, void 0, function* () {
        const response = yield client.images.generate({
            model: "dall-e-3",
            prompt: "A cute baby sea otter",
        });
        expect(response).toBeDefined();
        expect(response.data).toBeDefined();
        expect(response.data.length).toBeGreaterThan(0);
    }), 120000);
    test("createVariation: only required params", () => __awaiter(void 0, void 0, void 0, function* () {
        const imagePath = path_1.default.join(__dirname, 'image.png');
        const response = yield client.images.createVariation({
            image: fs_1.default.createReadStream(imagePath),
        });
        expect(response).toBeDefined();
        expect(response.data).toBeDefined();
        expect(response.data.length).toBeGreaterThan(0);
    }), 120000);
    test("edit: only required params", () => __awaiter(void 0, void 0, void 0, function* () {
        const imagePath = path_1.default.join(__dirname, 'image.png');
        const imageMaskPath = path_1.default.join(__dirname, 'imageMask.png');
        const response = yield client.images.edit({
            image: fs_1.default.createReadStream(imagePath),
            mask: fs_1.default.createReadStream(imageMaskPath),
            prompt: "A cute baby sea otter wearing a beret"
        });
        expect(response).toBeDefined();
        expect(response.data).toBeDefined();
        expect(response.data.length).toBeGreaterThan(0);
    }), 120000);
});
