// @empiricalrun/llm/dist/vision/captcha/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.solveRecaptchaGrid = void 0;
const __1 = require("../..");
const utils_1 = require("../utils");
async function solveRecaptchaGrid(input) {
    const { captchaType: type, captchaObject: object, screenshotBase64 } = input;
    const llm = new __1.LLM({
        provider: "openai",
    });
    const prompt = type === "images"
        ? promptForCaptchaImages(object)
        : promptForCaptchaSquares(object);
    const llmResponse = await llm.createChatCompletion({
        messages: [
            {
                role: "system",
                content: prompt,
            },
            {
                role: "user",
                content: [
                    {
                        type: "image_url",
                        image_url: {
                            url: (0, utils_1.imageFormatForProvider)("openai", screenshotBase64),
                        },
                    },
                ],
            },
        ],
        model: "gpt-4o",
    });
    if (!llmResponse) {
        throw new Error("Failed to extract text from image");
    }
    const response = llmResponse.content;
    const numbers = response
        .split(",")
        .map((n) => parseInt(n.trim()))
        .filter((n) => !isNaN(n));
    return numbers;
}
exports.solveRecaptchaGrid = solveRecaptchaGrid;
function promptForCaptchaSquares(object) {
    return `
You are given an image with a 4x4 grid overlayed on it. This image has an object
that we are very interested in. The object is: ${object}.

You need to help me select all grid boxes that contain this object.

Return your answer in the form of numbers separated by commas. Start
counting from 1 from the top left corner and go from left to right. The grid
numbering system looks like:

1 2 3 4
5 6 7 8
9 10 11 12
13 14 15 16

As an example, if the matching images are top row 3rd and bottom row 1st image,
then the answer would be "3,13". If there are no matching images, return "none".

DO NOT include any explanation or other text in your response.
`;
}
function promptForCaptchaImages(object) {
    return `
You are given a composite image that has a 3x3 grid of images. Select all the images
that match this: Select all the images that have ${object}.

Return your answer in the form of numbers separated by commas. Start
counting from 1 from the top left corner and go from left to right.

1 2 3
4 5 6
7 8 9

As an example, if the matching images are top row center and bottom row left,
then the answer would be "2,7". If there are no matching images, return "none".

DO NOT include any explanation or other text in your response.
`;
}
