// @empiricalrun/llm/dist/trace/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.shutdownLangfuse = exports.flushAllTraces = exports.langfuseInstance = exports.isLangfuseEnabled = void 0;
const langfuse_1 = require("langfuse");
exports.isLangfuseEnabled = process.env.LANGFUSE_SK && process.env.LANGFUSE_PK;
exports.langfuseInstance = exports.isLangfuseEnabled
    ? new langfuse_1.Langfuse({
        secretKey: process.env.LANGFUSE_SK,
        publicKey: process.env.LANGFUSE_PK,
        baseUrl: "https://us.cloud.langfuse.com",
        flushAt: 1,
    })
    : undefined;
/**
 * Flush all traces to trace service.
 * This is useful for debugging purposes.
 *
 * Usage example:
 * process.on("exit", async () => await langfuse.flushAsync());
 * process.on("SIGINT", async () => await langfuse.flushAsync());
 * process.on("SIGTERM", async () => await langfuse.flushAsync());
 */
async function flushAllTraces() {
    if (!exports.langfuseInstance) {
        console.warn("Langfuse instance is not initialized");
        return;
    }
    await exports.langfuseInstance?.flushAsync();
}
exports.flushAllTraces = flushAllTraces;
async function shutdownLangfuse() {
    if (!exports.langfuseInstance) {
        console.warn("Langfuse instance is not initialized");
        return;
    }
    await exports.langfuseInstance?.shutdownAsync();
}
exports.shutdownLangfuse = shutdownLangfuse;
